packages/grid/x-data-grid/src/hooks/features/columns/useGridColumnSpanning.ts

```diff
@@ -0,0 +1,142 @@
+import React from 'react';
+import { useGridApiMethod } from '../../utils/useGridApiMethod';
+import { GridColumnIndex, GridCellColSpanInfo } from '../../../models/gridColumnSpanning';
+import { GridRowId } from '../../../models/gridRows';
+import { GridColumnSpanning } from '../../../models/api/gridColumnSpanning';
+import { useGridApiEventHandler } from '../../utils/useGridApiEventHandler';
+import { GridEvents } from '../../../models/events/gridEvents';
+import { GridCellParams } from '../../../models/params/gridCellParams';
+import { GridStateColDef } from '../../../models/colDef/gridColDef';
+import { GridApiCommunity } from '../../../models/api/gridApiCommunity';
+
+/**
+ * @requires useGridColumns (method, event)
+ * @requires useGridParamsApi (method)
+ */
+export const useGridColumnSpanning = (apiRef: React.MutableRefObject<GridApiCommunity>) => {
+  const lookup = React.useRef<Record<GridRowId, Record<GridColumnIndex, GridCellColSpanInfo>>>({});
+
+  const setCellColSpanInfo = React.useCallback(
+    (rowId: GridRowId, columnIndex: GridColumnIndex, cellColSpanInfo: GridCellColSpanInfo) => {
+      const sizes = lookup.current;
+      if (!sizes[rowId]) {
+        sizes[rowId] = {};
+      }
+
+      sizes[rowId][columnIndex] = cellColSpanInfo;
+    },
+    [],
+  );
+
+  const getCellColSpanInfo = React.useCallback<GridColumnSpanning['unstable_getCellColSpanInfo']>(
+    (rowId, columnIndex) => {
+      return lookup.current[rowId]?.[columnIndex];
+    },
+    [],
+  );
+
+  // Calculate `colSpan` for the cell.
+  const calculateCellColSpan = React.useCallback(
+    (params: {
+      columnIndex: number;
+      rowId: GridRowId;
+      cellParams: GridCellParams;
+      renderedColumns: GridStateColDef[];
+    }) => {
+      const { columnIndex, rowId, cellParams, renderedColumns } = params;
+      const visibleColumns = apiRef.current.getVisibleColumns();
+      const columnsLength = visibleColumns.length;
+      const column = visibleColumns[columnIndex];
+
+      let width = column.computedWidth;
+
+      let colSpan =
+        typeof column.colSpan === 'function' ? column.colSpan(cellParams) : column.colSpan;
```

Comment:
I use it in `calculateColSpan` one level above https://github.com/cherniavskii/material-ui-x/blob/column-spanning/packages/grid/x-data-grid/src/hooks/features/columns/useGridColumnSpanning.ts#L119
To avoid passing `cellParams` to `calculateCellColSpan` I would need to pass `column.field` - I'm not sure that's a big improvement

Reply:
Right!
This also allowed to call `getCellParams` only when `colSpan` is a function and not call it otherwise 🎉